File: examples/rust/hand/agg/src/handlers/player_action.rs
//! PlayerAction command handler.

use angzarr_client::proto::examples::{ActionTaken, ActionType, PlayerAction};
use angzarr_client::proto::{CommandBook, EventBook};
use angzarr_client::{new_event_book, pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::{HandState, PlayerHandState};

/// Validated action result containing computed values.
struct ValidatedAction {
    chips_put_in: i64,
    event_amount: i64,
    final_action: ActionType,
}

fn guard(state: &HandState) -> CommandResult<()> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Hand does not exist"));
    }
    if state.is_complete() {
        return Err(CommandRejectedError::new("Hand already complete"));
    }
    if state.status != "betting" {
        return Err(CommandRejectedError::new("Not in betting phase"));
    }
    Ok(())
}

fn validate<'a>(
    cmd: &PlayerAction,
    state: &'a HandState,
) -> CommandResult<(&'a PlayerHandState, ValidatedAction)> {
    let player = state
        .get_player(&cmd.player_root)
        .ok_or_else(|| CommandRejectedError::new("Player not in hand"))?;

    if player.has_folded {
        return Err(CommandRejectedError::new("Player has folded"));
    }
    if player.is_all_in {
        return Err(CommandRejectedError::new("Player is all-in"));
    }

    let action = ActionType::try_from(cmd.action).unwrap_or_default();
    let amount_to_call = state.current_bet - player.bet_this_round;
    let chips_put_in: i64;
    let event_amount: i64;

    match action {
        ActionType::Fold => {
            chips_put_in = 0;
            event_amount = 0;
        }
        ActionType::Check => {
            if amount_to_call > 0 {
                return Err(CommandRejectedError::new("Cannot check, must call or fold"));
            }
            chips_put_in = 0;
            event_amount = 0;
        }
        ActionType::Call => {
            if amount_to_call <= 0 {
                return Err(CommandRejectedError::new("Nothing to call"));
            }
            chips_put_in = amount_to_call.min(player.stack);
            event_amount = chips_put_in;
        }
        ActionType::Bet => {
            if state.current_bet > 0 {
                return Err(CommandRejectedError::new("Cannot bet, use raise"));
            }
            if cmd.amount < state.min_raise {
                return Err(CommandRejectedError::new(format!(
                    "Bet must be at least {}",
                    state.min_raise
                )));
            }
            chips_put_in = cmd.amount.min(player.stack);
            event_amount = chips_put_in;
        }
        ActionType::Raise => {
            if state.current_bet <= 0 {
                return Err(CommandRejectedError::new("Cannot raise, use bet"));
            }
            let raise_amount = cmd.amount - state.current_bet;
            if raise_amount < state.min_raise {
                return Err(CommandRejectedError::new("Raise below minimum"));
            }
            let to_put_in = cmd.amount - player.bet_this_round;
            chips_put_in = to_put_in.min(player.stack);
            event_amount = cmd.amount;
        }
        ActionType::AllIn => {
            chips_put_in = player.stack;
            event_amount = chips_put_in;
        }
        _ => {
            return Err(CommandRejectedError::new("Unknown action"));
        }
    }

    let final_action = if chips_put_in == player.stack && chips_put_in > 0 {
        ActionType::AllIn
    } else {
        action
    };

    Ok((
        player,
        ValidatedAction {
            chips_put_in,
            event_amount,
            final_action,
        },
    ))
}

fn compute(
    cmd: &PlayerAction,
    state: &HandState,
    player: &PlayerHandState,
    validated: &ValidatedAction,
) -> ActionTaken {
    let new_stack = player.stack - validated.chips_put_in;
    let new_pot = state.total_pot() + validated.chips_put_in;
    let player_total_bet = player.bet_this_round + validated.chips_put_in;
    let new_current_bet = state.current_bet.max(player_total_bet);

    ActionTaken {
        player_root: cmd.player_root.clone(),
        action: validated.final_action as i32,
        amount: validated.event_amount,
        player_stack: new_stack,
        pot_total: new_pot,
        amount_to_call: new_current_bet,
        action_at: Some(angzarr_client::now()),
    }
}

pub fn handle_player_action(
    command_book: &CommandBook,
    command_any: &Any,
    state: &HandState,
    seq: u32,
) -> CommandResult<EventBook> {
    let cmd: PlayerAction = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    guard(state)?;
    let (player, validated) = validate(&cmd, state)?;

    let event = compute(&cmd, state, player, &validated);
    let event_any = pack_event(&event, "examples.ActionTaken");

    Ok(new_event_book(command_book, seq, event_any))
}

File: examples/go/hand/agg/handlers/player_action.go
package handlers

import (
	"fmt"
	"time"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

func guardPlayerAction(state HandState) error {
	if !state.Exists() {
		return angzarr.NewCommandRejectedError("Hand does not exist")
	}
	if state.IsComplete() {
		return angzarr.NewCommandRejectedError("Hand already complete")
	}
	if state.Status != "betting" {
		return angzarr.NewCommandRejectedError("Not in betting phase")
	}
	return nil
}

type validatedAction struct {
	player       *PlayerHandState
	actualAmount int64
}

func validatePlayerAction(cmd *examples.PlayerAction, state HandState) (*validatedAction, error) {
	player := state.GetPlayerByRoot(cmd.PlayerRoot)
	if player == nil {
		return nil, angzarr.NewCommandRejectedError("Player not in hand")
	}
	if player.HasFolded {
		return nil, angzarr.NewCommandRejectedError("Player has folded")
	}
	if player.IsAllIn {
		return nil, angzarr.NewCommandRejectedError("Player is all-in")
	}

	amountToCall := state.CurrentBet - player.BetThisRound
	actualAmount := int64(0)

	switch cmd.Action {
	case examples.ActionType_FOLD:
		// Always valid

	case examples.ActionType_CHECK:
		if amountToCall > 0 {
			return nil, angzarr.NewCommandRejectedError("Cannot check, must call or fold")
		}

	case examples.ActionType_CALL:
		if amountToCall <= 0 {
			return nil, angzarr.NewCommandRejectedError("Nothing to call")
		}
		actualAmount = amountToCall
		if actualAmount > player.Stack {
			actualAmount = player.Stack
		}

	case examples.ActionType_BET:
		if state.CurrentBet > 0 {
			return nil, angzarr.NewCommandRejectedError("Cannot bet, use raise")
		}
		minBet := state.MinRaise
		if minBet == 0 {
			minBet = 10
		}
		if cmd.Amount < minBet {
			return nil, angzarr.NewCommandRejectedError(fmt.Sprintf("Bet must be at least %d", minBet))
		}
		actualAmount = cmd.Amount
		if actualAmount > player.Stack {
			actualAmount = player.Stack
		}

	case examples.ActionType_RAISE:
		if state.CurrentBet <= 0 {
			return nil, angzarr.NewCommandRejectedError("Cannot raise, use bet")
		}
		totalBet := cmd.Amount
		raiseAmount := totalBet - state.CurrentBet
		if raiseAmount < state.MinRaise {
			return nil, angzarr.NewCommandRejectedError("Raise below minimum")
		}
		actualAmount = totalBet - player.BetThisRound
		if actualAmount > player.Stack {
			actualAmount = player.Stack
		}

	case examples.ActionType_ALL_IN:
		actualAmount = player.Stack

	default:
		return nil, angzarr.NewCommandRejectedError("Unknown action")
	}

	return &validatedAction{player: player, actualAmount: actualAmount}, nil
}

func computeActionTaken(cmd *examples.PlayerAction, state HandState, va *validatedAction) *examples.ActionTaken {
	newStack := va.player.Stack - va.actualAmount
	newPot := state.TotalPot() + va.actualAmount

	newCurrentBet := state.CurrentBet
	playerTotalBet := va.player.BetThisRound + va.actualAmount
	if playerTotalBet > newCurrentBet {
		newCurrentBet = playerTotalBet
	}

	action := cmd.Action
	if va.actualAmount == va.player.Stack && va.actualAmount > 0 {
		action = examples.ActionType_ALL_IN
	}

	amountToEmit := va.actualAmount
	if cmd.Action == examples.ActionType_BET || cmd.Action == examples.ActionType_RAISE {
		amountToEmit = cmd.Amount
	}

	return &examples.ActionTaken{
		PlayerRoot:   cmd.PlayerRoot,
		Action:       action,
		Amount:       amountToEmit,
		PlayerStack:  newStack,
		PotTotal:     newPot,
		AmountToCall: newCurrentBet,
		ActionAt:     timestamppb.New(time.Now()),
	}
}

// HandlePlayerAction handles the PlayerAction command.
func HandlePlayerAction(
	commandBook *pb.CommandBook,
	commandAny *anypb.Any,
	state HandState,
	seq uint32,
) (*pb.EventBook, error) {
	var cmd examples.PlayerAction
	if err := proto.Unmarshal(commandAny.Value, &cmd); err != nil {
		return nil, err
	}

	if err := guardPlayerAction(state); err != nil {
		return nil, err
	}
	va, err := validatePlayerAction(&cmd, state)
	if err != nil {
		return nil, err
	}

	event := computeActionTaken(&cmd, state, va)

	eventAny, err := anypb.New(event)
	if err != nil {
		return nil, err
	}

	return angzarr.NewEventBook(commandBook.Cover, seq, eventAny), nil
}
